Add a certificate renewal script

bickelj opened the pull request:
When run by cron, this script automatically renews certificates using certbot for two hostnames and sends a chat message with the result.

Issue #64 Renew TLS certificates

jasonaowen commented on renewCerts.sh:
```diff
@@ -0,0 +1,73 @@
+    exit $exit_code
+}
+
+certbot certonly -n --domain=$AUTH_SERVER_HOSTNAME --standalone --keep-until-expiring \
```
How does the `--standalone` option ("Run a standalone webserver for authentication") work with the existing port mapping? I would have expected that to fail!

bickelj replied:
In order for this to work, we do not listen on port 80 on the reverse proxy. Certbot does its hostname validation via port 80. After all certbot is creating a certificate for use on port 443 so to require 443 would cause deadlock. Not listening on port 80 means we unfortunately can't redirect plain http traffic on port 80 to https on 443. Then again it is 2023 so we've had a decade or more to start defaulting to `https` when filling out web addresses.

jasonaowen replied:
Ah. That is a limitation, but at least I did understand all the moving parts correctly! Thanks for the explanation, @bickelj.

jasonaowen commented on renewCerts.sh:
```diff
@@ -0,0 +1,73 @@
+certbot certonly -n --domain=$WEB_SERVER_HOSTNAME --standalone --keep-until-expiring \
+    || fin $(tail -n 10 /var/log/letsencrypt/letsencrypt.log)
+# Make sure the reverse proxy user has read/execute privileges to the files
+chown -R root:$REVERSE_PROXY_CONTAINER_USER /etc/letsencrypt/{live,archive} \
```
If we're leaving the ownership alone and just changing the group, let's rename the variable to `REVERSE_PROXY_CONTAINER_GROUP` and call `chgrp(1)` instead:

```suggestion
chgrp -R $REVERSE_PROXY_CONTAINER_GROUP /etc/letsencrypt/{live,archive} \
```

bickelj replied:
+1 to the `chgrp` but we use the `REVERSE_PROXY_CONTAINER_USER` as the user that runs `nginx`. The idea is to give read access to that `nginx` user and nobody else.

bickelj replied:
Reading between the lines, though, there is no guarantee that the `gid` matches the `uid`, so we should separate `GROUP` from `USER`.

jasonaowen commented on renewCerts.sh:
```diff
@@ -0,0 +1,73 @@
+set -eo pipefail
+
+test -x "$(which certbot)"
+. /home/deploy/.env
```
Love the `test`s below for environment variables, but not a fan of sourcing a `.env` file, as we've previously discussed. More importantly, this script is being run as root (right?), so sourcing a file owned by a non-root user is a security vulnerability: an attacker could compromise the `deploy` user, and then this would be an escalation path to compromise the entire machine. If we must have a `.env` file, it should live in `~root`.

bickelj replied:
Yes, good points all around.

bickelj replied:
Maybe this environment variables discussion warrants another ticket because there are several things to unpack. For the purpose of this change, the `deploy` user has super-user privileges by virtue of `docker` group membership. If the `deploy` user gets compromised the machine is already lost. You had raised a concern about the web service component reading a `.env` file as distinct from the use of environment variables at all, which I agree with: the components should probably not try to read any `.env` files but rather get them from the process, the environment, as configured by the administrator. But at some point those environment variables need to be set. Perhaps there is some better or safer way to store the environment variables besides a `.env` file, but this still only pushes the problem back one step because that mechanism will still need access to a secret somewhere, probably also in a file (even if indirectly).

At the moment what I can do is ensure that `/home/deploy` and especially `.env` can only be read by the `deploy` user. Perhaps there is also a way to get the environment variable values without sourcing the entire file. I am open to solutions.

jasonaowen replied:
> Maybe this environment variables discussion warrants another ticket because there are several things to unpack.

Sure, we don't have to hash this all out here!

> For the purpose of this change, the `deploy` user has super-user privileges by virtue of `docker` group membership. If the `deploy` user gets compromised the machine is already lost.

What do you mean by this? Users in the `docker` group have blanket no-password `sudo` privileges? That would be concerning!

If not, if the group just grants access to docker-y things, then I think that's a small but important distinction: an attacker would be able to compromise the virtual machines being run - which are, indeed, the vast majority of the value of the machine - but still not the host machine. Assuming I'm understanding correctly that the script is intended to be run as the root user - for example, by putting a copy in `/etc/cron.weekly` - then this is still an additional escalation path that we ought not introduce.

> You had raised a concern about the web service component reading a `.env` file as distinct from the use of environment variables at all, which I agree with: the components should probably not try to read any `.env` files but rather get them from the process, the environment, as configured by the administrator. But at some point those environment variables need to be set. Perhaps there is some better or safer way to store the environment variables besides a `.env` file, but this still only pushes the problem back one step because that mechanism will still need access to a secret somewhere, probably also in a file (even if indirectly).

Systemd has the [`EnvironmentFile=` directive](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=), which alongside standard *nix file permissions should be a secure method of storing and loading secrets.

But yes, +1 to moving that discussion out of this PR; it's way out of scope!

> At the moment what I can do is ensure that `/home/deploy` and especially `.env` can only be read by the `deploy` user. Perhaps there is also a way to get the environment variable values without sourcing the entire file. I am open to solutions.

Well, my original suggestion is to move the environment file used by this script from the `deploy` user's home directory to `root`'s. Again, that's predicated on my assumption that this script is being executed by the `root` user, which I based on the lack of calls to `sudo` around invocations of `certbot` and `chown` - if it is actually being executed by `deploy`, then sourcing a file in `deploy`'s homedir adds no new escalation path to `root`!

jasonaowen replied:
> Systemd has the [`EnvironmentFile=` directive](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=), which alongside standard *nix file permissions should be a secure method of storing and loading secrets.

Systemd also has a more secure [credential system](https://systemd.io/CREDENTIALS/), but I believe that requires special application support, and at least in this case I don't think that level of security is worth the additional complexity. But, I wanted to mention it for the sake of completeness!

bickelj replied:
> What do you mean by [the `deploy` user has super-user privileges by virtue of `docker` group membership]? Users in the docker group have blanket no-password sudo privileges? That would be concerning!

The `deploy` user does docker-y things as you suggested: 
1. [`docker compose ... pull ...` (get the new software)](https://github.com/PhilanthropyDataCommons/deploy/blob/28de23d3ee004ff26a5eb76e5d5646f531781dee/deploy.sh#L108),
2. [`docker compose ... down ... ` (stop the old software)](https://github.com/PhilanthropyDataCommons/deploy/blob/28de23d3ee004ff26a5eb76e5d5646f531781dee/deploy.sh#L111),
3. [`docker compose ... up ...` (start the new software)](https://github.com/PhilanthropyDataCommons/deploy/blob/28de23d3ee004ff26a5eb76e5d5646f531781dee/deploy.sh#L119).

In the default docker configuration, in order to run `docker [command]`, one must be root, be able to sudo to root, or be a member of the docker group. One of those commands is `run` which runs a container based on an image. Many images run as the root user and therefore if you can `docker run` you can run an image as root, thus privilege escalation, thus "the `deploy` user has super-user privileges by virtue of `docker` group membership."

So yes, one of the concerns with the default docker setup is the privileges required or gained by running docker commands. RedHat's answer is [podman](https://www.redhat.com/en/topics/containers/what-is-podman), [buildah](https://www.redhat.com/en/topics/containers/what-is-buildah), and [skopeo](https://www.redhat.com/en/topics/containers/what-is-skopeo). I am not necessarily suggesting a switch to those solutions because I am sure they come with their own baggage, but highlighting that your concern is known and valid.

Back to our setup, the `deploy` user is not able to log in, and other users should not be added to the `deploy` group. The reason for using `/home` instead of other directories is to clearly separate the docker-y data directories from host system directories. There may be a cleaner way to achieve that separation and I'm open to that. I wouldn't want to try to use the system directories from docker, though, because if someone does `apt install postgresql` that command should work without conflict. The `.env` file is primarily for `docker compose` to set up environments for each of the containers. That is why it lives in `/home/deploy`. I can try putting it in `/root` and tell `docker compose` to look there explicitly rather than the current setup where `docker compose` implicitly looks for a `.env` file in the same directory where the command runs.

Another option besides a shared single `.env` file used both by this script that runs as root and the `deploy` user's `.env` file is to suffer some duplication and have a separate one for this script in `/root`, I think that would probably be fine too.

And yet another option would be to specify the hostnames directly in the crontab such that no `.env` file is required. This could be done as args to the script too.

bickelj replied:
`docker-compose` fails to read from `/root/.env` (not a bad thing):
> docker-compose --env-file /root/.env -f compose-dev.yml up
> WARNING: The REVERSE_PROXY_CONTAINER_USER variable is not set. Defaulting to a blank string.
> ... [similar WARNINGs for the rest]

So we either need to duplicate env vars or mitigate the way we read them. This seemed to work fine and should mitigate some of the attacks:
```
...
# Instead of sourcing /home/deploy/.env verbatim, source only strictly formed and used env vars.
. <(egrep '^(WEB_SERVER_HOSTNAME|REVERSE_PROXY_CONTAINER_USER|AUTH_SERVER_HOSTNAME|ZULIP_BASE_URL|ZULIP_BOT_EMAIL_ADDRESS|ZULIP_BOT_API_KEY|ZULIP_STREAM|ZULIP_TOPIC)=[a-zA-Z0-9"\/\:\.\@\_\-]+$' /home/deploy/.env)
```
The idea being that if we only get exactly the env vars we need, there are no spaces or `$` or parens or redirects in the values, and at the end of the value is a newline, we should be less likely to have arbitrary commands run from sourcing these env vars.

I am really trying to avoid duplicating these env vars, as you can tell!

jasonaowen replied:
This seems like a reasonable local maximum - thanks for following up on this, @bickelj!

jasonaowen replied:
I don't feel like I gave this solution the praise it deserved - limiting the scope via grep with a fine-grained regex is a good idea! Well done @bickelj.

jasonaowen reviewed (changes requested):
Thanks for working on this, @bickelj! I'll be very glad to have this automated.

The amount of work it requires is concerning; it is surprising to me how difficult something so basic and so critical is! I don't have a quick and easy alternative to suggest, though, and I'm certainly not suggesting we spend time on revamping our deployment architecture at this time.

I haven't tested this at all, but I've seen some Zulip comments, so presumably this is working!

A comment, a suggestion, and a requested change, below.

bickelj committed:
Remove an unused env var

This was found when looking to add another env var.

Issue #64 Renew TLS certificates

bickelj committed:
Add a certificate renewal script

When run by cron, `renewCerts.sh` automatically renews certificates
using certbot for `WEB_SERVER_HOSTNAME` and `AUTH_SERVER_HOSTNAME` and
sends a chat message with the result.

It is important that the user running nginx can read the certificates
and keys. To ensure the same group is used for both the containers and
this script, source env vars from the `deploy` user's `.env` file. To
mitigate the possibility of commands being run from the `.env` file,
require strict formatting, e.g. 'EXPECTED_VAR=some-value@place.com' as
the full line, with an allowed list of names like `EXPECTED_VAR`.

Before updating the `WEB_CERT`, `WEB_KEY`, `AUTH_CERT`, `AUTH_KEY` env
vars, run the `renewCerts.sh` at least once on the host. This way the
`/etc/letsencrypt` directories can be populated before attempting to
read them from the containers.

A new `REVERSE_PROXY_CONTAINER_GROUP` variable is needed. The value
should usually match that of the `REVERSE_PROXY_CONTAINER_USER` but it
may not always. Check `/etc/groups` to see what it is before setting.

Issue #64 Renew TLS certificates

jasonaowen reviewed (approved):
Thanks, @bickelj!
